src/igraph/__init__.py

```diff
@@ -3190,6 +3190,54 @@ def DataFrame(klass, edges, directed=True, vertices=None, use_vids=False):
 
         return g
 
+    def get_vertex_dataframe(self):
+        """Export vertices with attributes to pandas.DataFrame
+
+        @return: a pandas.DataFrame representing vertices and their attributes.
+          If the vertices have names, the index of the dataframe contains them,
+          otherwise it's integers from 0 up (which are also the vertex IDs).
+          The order of the vertices is the same as in the graph, so the n-th
+          row describes the vertex with ID i.
+        """
+        import pandas as pd
+
+        df = pd.DataFrame(
+            {attr: self.vs[attr] for attr in self.vertex_attributes()},
+            index=list(range(self.vcount())),
+        )
+        if 'name' in df.columns:
+            df.set_index('name', inplace=True)
+        return df
+
+    def get_edge_dataframe(self):
+        """Export edges with attributes to pandas.DataFrame
+
+        @return: a pandas.DataFrame representing edges and their attributes.
+          If the edges have names, the index of the dataframe contains them,
+          otherwise it's integers from 0 up (which are also the edge IDs).
+          The order of the edges is the same as in the graph, so the n-th
+          row describes the edge with ID i. The first two columns of the
+          dataframe represent the IDs of source and target vertices for each
+          edge. These columns have names "source" and "target". If your
+          edges have attributes with the same names, they will be present in
+          the dataframe, but not in the first two columns.
+        """
+        import pandas as pd
+
+        df = pd.DataFrame(
+            {attr: self.es[attr] for attr in self.edge_attributes()},
+            index=list(range(self.ecount())),
+        )
+        if 'name' in df.columns:
+            df.set_index('name', inplace=True)
+
+        df.insert(
+            0, 'source', [e.source for e in self.es], allow_duplicates=True)
```

Comment:
It might make sense here to use the `name` of the `node_df`, because that is effectively also the `index` of the `vertex_df`.

Reply:
This is a good point, that related to the https://github.com/igraph/python-igraph/pull/355#discussion_r558615150 below (why on Earth did I start replying there!)

Whichever we choose I agree the `source` and `target` cols/index levels should have the same thing as used to index the vertex dataframe, so that's not consistent at the moment.

I'd be tempted to advocate using vertex ids everywhere (i.e. keep this one and revert the vertex df to an index with just ids) because we know for a fact they are unique and well formatted. If we at some point start allowing people to use any Python hashables as vertex names, they will need to be here too which I'm a little afraid might cause a mess with dtypes.

But as for this whole PR, I'm only slightly opinionated, if you feel like you are using vertex names a lot more often than ids I'm happy to switch over everything to vertex names.

Reply:
> If we at some point start allowing people to use any Python hashables as vertex names,

oh please don't even think about that :) One has to draw a line somewhere.

Reply:
We could do: anything that can be converted into a hashable without a loss of generality is accepted 😉

Reply:
Ha, I actually use this somewhere: I have a graph that determines how other graphs are coupled, see https://leidenalg.readthedocs.io/en/stable/multiplex.html#slices-to-layers.

But using such graphs as indices might become a bit more complicated for a pandas dataframe. In principle, pandas dataframes can simply store any Python object, as far as I know. I am not sure if there are any limitations on this for indices though?

I think it would make more sense to start working with names as much as possible (or at least try to support that if people want to), so for me, having this as an option would be good. Of course, it should somehow be consistent, but for that reason I think it should be left to a user to specify whether he wants the index to be with names, or with vertex integer identifiers. The default option should then be consistent with each other (and probably default to integers, or perhaps default to names only if names are available?)